### DISCUSSION TITLE
Cannot create multiple errors response when generate Document

### DISCUSSION THREAD

**Discussion by huyhoang160593** [RobinTail/express-zod-api#1332] [Q&A] [answered]
So the thing it in the doc there are mention to create custom response here https://github.com/RobinTail/express-zod-api#response-customization. But you can only create 1 schema for 1 negative response, but if I want to create multiple errors like 400 for validate and 401 for authorization, there is no way to define that. How can I utilize the custom response to have multiple entries of error in the same result handler?

**Comment by RobinTail** [comment]
Hello @huyhoang160593 
The cause for a negative response is an `Error`.
Every error has a reason, which is is supposed to be expressed in its `message` property.
Surely, there can be custom errors that extend `Error` with additional properties and metadata.
You can create a schema for a negative response that covers those scenarios.

Consider [discriminated union schema](https://github.com/colinhacks/zod#discriminated-unions) for that purpose:

```ts
const negativeResponseSchema = z.discriminatedUnion("cause", [
  z.object({ cause: z.literal("validation"), message: z.string() }),
  z.object({ cause: z.literal("authorization"), details: z.object({ ...}) }),
]);
```

However, the generated documentation currently is designed for a single negative schema that does not depend on particular faulty status code. It's purpose was to illustrate what the negative payload could be, but not to differentiate between `400` and `401`.

This could probably be improved later, but so far a nice way is not yet invented.

**Comment by RobinTail** [comment]
@huyhoang160593 , would you be interested in helping me to test the feature #1431 ?

**Reply by huyhoang160593** [reply]
@RobinTail I would love to but the things is that I don't know how to use your package as a local package, after building the branch, I have tried to link the package to my mini reproduction project and run it without any success and have this as return.
```
TypeError: Cannot read properties of undefined (reading '_def')
    at Le (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:18:9089)
    at Le (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:18:9084)
    at Er (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:18:9288)
    at onEndpoint (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:18:14895)
    at <anonymous> (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:5:8984)
    at Array.forEach (<anonymous>)
    at <anonymous> (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:5:8846)
    at Array.forEach (<anonymous>)
    at ce (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:5:8481)
    at <anonymous> (/home/the99spuppycat/Desktop/temp/zod-express-mini-production/local_package/express-zod-api/dist/index.cjs:5:9134)
```
Can you help me with the linking local package so I can properly testing.
The step I have done:
1. clone your repo with branch[better-status-codes2](https://github.com/RobinTail/express-zod-api/tree/better-status-codes2)
2. run `yarn install` to install all dependency
3. run `yarn build` and it run successfully
4. link the repo to my project with yarn add <relative-path-to-the-repo> and my `package.json` display that I have link successfully with the name `zod-express-api` and the version is `express-zod-api@16.2.0-beta1`
5. run my project and the error appear.

**Reply by RobinTail** [reply]
> I would love to but the things is that I don't know how to use your package as a local package

You don't have to do that, @huyhoang160593 .
For testing purposes I published the version `16.2.0-beta1`.
So, instead of linking it locally, just install it from NPM using dist tag `beta`:

```sh
yarn add express-zod-api@beta
```

or adjust the version explicitly in your `package.json`:

```json lines
"dependencies": {
  "express-zod-api": "^16.2.0-beta1"
}
```

and run `yarn install`.

**Reply by RobinTail** [reply]
When you finished installation, @huyhoang160593 , use the featured syntax for making a result handler according to your needs.
Please test how it responds and how it generates the documentation of your API.

This is an example factory with a features kind of ResultHandler:
https://github.com/RobinTail/express-zod-api/blob/749fafc3b284980899ea7d521a9ac142943b91de/example/factories.ts#L91

And this is an example endpoint produced on that factory:
https://github.com/RobinTail/express-zod-api/blob/better-status-codes2/example/endpoints/create-user.ts

More details in the Changelog: https://github.com/RobinTail/express-zod-api/blob/better-status-codes2/CHANGELOG.md#1620

Please let me know the results of your test including the `ResultHandler` you made.
In case of confusion or unclarity or lack of documentation, please let me know as well.
In case the error you mentioned above appears again, please publish your repo on GitHub and give me a link, so I could debug it myself.
Thank you

**Reply by huyhoang160593** [reply]
@RobinTail the document seem like working just fine, but the typescript is not so happy about the code and all my custom `ResultHandler` is warning and red like the end of the earth may occur. This is the link to my reproduction repo, please give a look and check [here](https://github.com/huyhoang160593/express-zod-api-mini-repoduction)

**Reply by RobinTail** [reply]
> but the typescript is not so happy about the code

I checked it, @huyhoang160593 
This is because you have mixed syntax in your Result Handler:
- Your `getNegativeResponse` returns an array (the featured syntax),
- While your `getPositiveResponse` (which appears to me a bit overcomplicated) returns different schemas (not array, previous syntax).

Those syntaxes can not be mixed, because [overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) was the only way to introduce the feature.

So, in order to resolve the issue, those methods should both return EITHER arrays OR not arrays.
When I wrapped each of your `return` statements of `getPositiveResponse` with `[{ schema: ... }]`, that fixed the issue:

![image](https://github.com/RobinTail/express-zod-api/assets/13189514/0bb9a711-e634-44fe-8536-29e3f9598337)

Try it yourself and thank you for informing me.

**Reply by RobinTail** [reply]
Good news, @huyhoang160593 .
I think I found a way to improve my implementation to enable mixing like in your case.

**Reply by RobinTail** [reply]
@huyhoang160593 , I published version `16.2.0-beta2`, please try it with your code as is — it should work.
Now it checks programmatically if the array of possible response schemas is not empty.
Please let me know how it works. Thank you

**Reply by huyhoang160593** [reply]
@RobinTail omg thank you for the hard work, I'm very appreciate it, this work wonderfully in my repo now

**Reply by RobinTail** [reply]
Great, I'm glad to read it, @huyhoang160593 .
Thank you as well for participating in the testing.
I could not came up with such use cases myself.
I'm going to await for one more test of this version by [Dan Clayton in discussion #1193](https://github.com/RobinTail/express-zod-api/discussions/1193), and then make a stable release.

**Comment by RobinTail** [comment, marked as answer]
For those looking for a solution to the same problem,
the feature is available in version 16.2.0 🚀 — see [Changelog for details](https://github.com/RobinTail/express-zod-api/blob/master/CHANGELOG.md#1620).
